driver-core/src/main/java/com/datastax/driver/core/Fetchable.java

```diff
@@ -0,0 +1,166 @@
+/*
+ *      Copyright (C) 2012-2015 DataStax Inc.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+package com.datastax.driver.core;
+
+import com.google.common.util.concurrent.ListenableFuture;
+
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * Defines a result set whose rows can be fetched and paged,
+ * possibly asynchronously.
+ */
+public interface Fetchable<S extends Fetchable<S, T>, T> extends Iterable<T> {
+
+    /**
+     * Returns whether this result set has more results.
+     *
+     * @return whether this result set has more results.
+     */
+    boolean isExhausted();
+
+    /**
+     * Whether all results from this result set have been fetched from the
+     * database.
+     * <p/>
+     * Note that if {@code isFullyFetched()}, then {@link #getAvailableWithoutFetching}
+     * will return how many rows remain in the result set before exhaustion. But
+     * please note that {@code !isFullyFetched()} never guarantees that the result set
+     * is not exhausted (you should call {@link #isExhausted()} to verify it).
+     *
+     * @return whether all results have been fetched.
+     */
+    boolean isFullyFetched();
+
+    /**
+     * The number of rows that can be retrieved from this result set without
+     * blocking to fetch.
+     *
+     * @return the number of rows readily available in this result set. If
+     * {@link #isFullyFetched()}, this is the total number of rows remaining
+     * in this result set (after which the result set will be exhausted).
+     */
+    int getAvailableWithoutFetching();
+
+    /**
+     * Force fetching the next page of results for this result set, if any.
+     * <p/>
+     * This method is entirely optional. It will be called automatically while
+     * the result set is consumed (through {@link #one}, {@link #all} or iteration)
+     * when needed (i.e. when {@code getAvailableWithoutFetching() == 0} and
+     * {@code isFullyFetched() == false}).
+     * <p/>
+     * You can however call this method manually to force the fetching of the
+     * next page of results. This can allow to prefetch results before they are
+     * strictly needed. For instance, if you want to prefetch the next page of
+     * results as soon as there is less than 100 rows readily available in this
+     * result set, you can do:
+     * <pre>
+     *   ResultSet rs = session.execute(...);
```

Comment:
The javadocs still mention `ResultSet` in some places but I guess it's ok since `ResultSet` is the only child interface of this one in this module.